runtime/interp.c

```diff
@@ -1372,7 +1374,9 @@ do_resume: {
       if (parent == NULL) {
         accu = caml_continuation_use(cont);
         resume_fn = raise_unhandled_effect;
+        Setup_for_gc;
         resume_arg = caml_make_unhandled_effect_exn(eff);
+        Restore_after_gc;
```

Comment:
I find `Setup_for_gc` around a function that is not an immediate GC call rather weird / unidiomatic, but I think that it is correct -- it is needed to preserve the `accu` global that is set on line 1375. Another option would be to use `Setup_for_c_call` instead, but move the `accu = ` on line 1375 after the allocating call, so that `accu` needs not be preserved.

Reply:
The other thing I considered - given that `caml_make_unhandled_effect_exn` is a refactoring which is only present to allow the bytecode interpreter to do this was either to convert it to a `Make_unhandled_effect_exn` macro (and so use `Alloc_small` with `Enter_gc` directly) or a `Caml_inline` function in `interp.c`.

Reply:
I thought about this as well, but it is also used by `caml_raise_unhandled_effect` which is itself used by runtime files through `caml_c_call`.

Reply:
What I meant was to merge `caml_raise_unhandled_effect` and `caml_make_unhandled_effect_exn` back together (for native code) and then hand-roll a bytecode version in interp.c - the allocation function differs in each case. It wasn't clear to me that `Setup_for_c_call` is necessarily the _right_ thing here.

Reply:
`Setup_for_gc` .. `Restore_after_gc` is used in `Enter_gc` - I think is a bracketting it's more "setup for the _possibility_ of the gc being called", i.e. it brackets something which allocates?

Reply:
Yes, but `Enter_gc` itself is only used in `Alloc_small` invocations and nowhere else. You are breaking this pattern by using it to call an apparently-arbitrary C function (this function is not special, many C functions can call the OCaml GC, including all the unknown functions in the `C_CALL<n>` instructions).

> What I meant was to merge `caml_raise_unhandled_effect` and `caml_make_unhandled_effect_exn` back together (for native code) and then hand-roll a bytecode version in interp.c - the allocation function differs in each case.

Yes, I think that would work. (Note: I'm slightly annoyed by `caml_raise(...)` taking a `value` temporary without rooting; if you rewrite this function I hope this will go away.)

Reply:
Hmm, I just had a look at rewriting the function, but it's really not as neat as I initially thought it might be - there's the exception caching, which either has to be exposed (slightly hairy, given the atomics, etc.) or _duplicated_.

This function is not special, but the function _call_ is special - as far as I can see, it's the only instance in the entire interpreter loop where we call a function which may allocate **and** where `accu` needs to be preserved across it. It looks to me (possibly incorrectly), that _all_ the other instances are in fact `Setup_for_gc` unless they can either be proven not to be needed at all (several comments at the start of the interpreter around use of `caml_alloc_shr`) or can be _optimised_ to `Setup_for_c_call` where we know that either `accu` is registered as a root by the function being called and so does not need to be preserved (C_CALL) cases or is not live/already registered (exception raising cases).

Reply:
There may be a simpler way to go about all this, let me think about it.  Meanwhile, I'm puzzled by the line
```
accu = caml_continuation_use(cont);
```
It looks like the value of `accu` is not used by the code at label `do_resume`.  Do you agree?  What does this mean? (to ignore the return value of `caml_continuation_use`)?